libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -0,0 +1,801 @@
+#include <uavcan/equipment/gnss/Auxiliary.hpp>
+#include <uavcan/equipment/ahrs/MagneticFieldStrength.hpp>
+#include <uavcan/equipment/air_data/StaticPressure.hpp>
+#include <uavcan/equipment/air_data/StaticTemperature.hpp>
+#include <uavcan/equipment/actuator/ArrayCommand.hpp>
+#include <uavcan/equipment/actuator/Command.hpp>
+#include <uavcan/equipment/actuator/Status.hpp>
+#include <uavcan/equipment/esc/RawCommand.hpp>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define debug_uavcan(level, fmt, args...) do { if ((level) <= AP_BoardConfig::get_can_debug()) { hal.console->printf(fmt, ##args); }} while (0)
+
+// Translation of all messages from UAVCAN structures into AP structures is done
+// in AP_UAVCAN and not in corresponding drivers.
+// The overhead of including definitions of DSDL is very high and it is best to
+// concentrate in one place.
+
+// TODO: temperature can come not only from baro. There should be separation on node ID
+// to check where it belongs to. If it is not baro that is the source, separate layer
+// of listeners/nodes should be added.
+
+// table of user settable CAN bus parameters
+const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
+    // @Param: NODE
+    // @DisplayName:  UAVCAN node that is used for Ardupilot
+    // @Description: UAVCAN node should be set implicitly
+    // @Range: 1 250
+    // @User: Advanced
+    AP_GROUPINFO("NODE", 1, AP_UAVCAN, _uavcan_node, 10)
+,
+
+AP_GROUPEND
+};
+
+static uavcan::Subscriber<uavcan::equipment::gnss::Fix> *gnss_fix;
+static void gnss_fix_cb(const uavcan::ReceivedDataStructure<uavcan::equipment::gnss::Fix>& msg)
+{
+    if (hal.can_mgr != nullptr) {
+        AP_UAVCAN *ap_uavcan = hal.can_mgr->get_UAVCAN();
+        if (ap_uavcan != nullptr) {
+            AP_GPS::GPS_State *state = ap_uavcan->find_gps_node(msg.getSrcNodeID().get());
+
+            if (state != nullptr) {
+                bool process = false;
+
+                if (msg.status == uavcan::equipment::gnss::Fix::STATUS_NO_FIX) {
+                    state->status = AP_GPS::GPS_Status::NO_FIX;
+                } else {
+                    if (msg.status == uavcan::equipment::gnss::Fix::STATUS_TIME_ONLY) {
+                        state->status = AP_GPS::GPS_Status::NO_FIX;
+                    } else if (msg.status == uavcan::equipment::gnss::Fix::STATUS_2D_FIX) {
+                        state->status = AP_GPS::GPS_Status::GPS_OK_FIX_2D;
+                        process = true;
+                    } else if (msg.status == uavcan::equipment::gnss::Fix::STATUS_3D_FIX) {
+                        state->status = AP_GPS::GPS_Status::GPS_OK_FIX_3D;
+                        process = true;
+                    }
+
+                    const uint64_t posix_offset = UNIX_OFFSET_MSEC;
+                    const uint64_t ms_per_week = 7ULL * 24 * 3600 * 1000;
+                    uint64_t epoch_ms = uavcan::UtcTime(msg.gnss_timestamp).toUSec();
```

Comment:
> no more reasons are needed to support @tridge's idea of having protocol (UART, I2C, SPI) encapsulation so we can use our own drivers.

This is hardly going to happen due to the reasons discussed in the chat room.

> after looking at Zubax source, it only sends time when it has an UTC valid time, which may take a long time...

The UTC time becomes immediately available once the unit captures at least one Glonass satellite, so the statement about the long time doesn't really hold. Despite this, shall it be desired, we can add support for GPS time instead of UTC time to Zubax GNSS.